### ISSUE TITLE
Add lint to detect unnecessary `clone()` for `AsRef`-taking function

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-clippy#12455]
### What it does

Consider this code:
```rust
use std::path::*;

fn item(_x: impl AsRef<Path>) {}

pub fn work(x: PathBuf) {
    item(x.clone()); // unnecessary clone
    item(&x);
}
```
This should be better written as
```rust
use std::path::*;

fn item(_x: impl AsRef<Path>) {}

pub fn work(x: PathBuf) {
    item(&x);
    item(&x);
}
```
Clippy as a bunch of lints against unnecessary borrowing, which is syntactic noise but ultimately usually harmless. Unnecessary cloning however can make the program slower for no good reason, so it is IMO quite relevant to lint against as well.

Clippy is already able to detect similar cases, like unnecessary `to_owned()`:
```rust
use std::path::*;

fn item(_x: impl AsRef<Path>) {}

pub fn work(x: &Path) {
    item(x.to_owned()); // clippy lints here
    item(x);
}
```

### Advantage

Same as the existing `unnecessary_to_owned`: avoids unnecessary copies and function calls, making the code both less noisy and faster.

### Drawbacks

If it can be implemented without false positivies, I cannot think of any drawbacks.

### Example

(see description above)
